src/android/adapter.rs

```diff
@@ -0,0 +1,316 @@
+use std::collections::HashMap;
+use std::pin::Pin;
+use std::sync::atomic::{AtomicI32, Ordering};
+use std::sync::{Arc, Mutex};
+use std::task::{Context, Poll};
+
+use futures_channel::mpsc::{Receiver, Sender};
+use futures_core::Stream;
+use futures_lite::{stream, StreamExt};
+use jni_glue::{Argument, ByteArray, Env, Global, Local, PrimitiveArray, VM};
+use tracing::{debug, warn};
+use uuid::Uuid;
+
+use super::bindings::android::bluetooth::le::{BluetoothLeScanner, ScanResult};
+use super::bindings::android::bluetooth::{BluetoothAdapter, BluetoothManager};
+use super::bindings::android::os::ParcelUuid;
+use super::bindings::com::github::alexmoon::bluest::android::BluestScanCallback;
+use super::device::DeviceImpl;
+use super::{JavaIterator, OptionExt};
+use crate::android::bindings::java::util::Map_Entry;
+use crate::util::defer;
+use crate::{
+    AdapterEvent, AdvertisementData, AdvertisingDevice, ConnectionEvent, Device, DeviceId, ManufacturerData, Result,
+};
+
+struct AdapterInner {
+    manager: Global<BluetoothManager>,
+    _adapter: Global<BluetoothAdapter>,
+    le_scanner: Global<BluetoothLeScanner>,
+}
+
+#[derive(Clone)]
+pub struct AdapterImpl {
+    inner: Arc<AdapterInner>,
+}
+
+impl AdapterImpl {
+    pub unsafe fn new(vm: *mut jni_glue::sys::JavaVM, manager: jni_glue::sys::jobject) -> Result<Self> {
+        let vm = VM::from_raw(vm);
+        let manager: Global<BluetoothManager> = Global::from_raw(vm, manager);
```

Comment:
Could this just take a `VM` and a `Global<BluetoothManager>` instead of the raw pointers? I think that would push the unsafe bits onto the invariants of `VM::from_raw` and `Global::from_raw` and this could become safe?

Reply:
I didn't do it because it'd require making `crate::android::bindings` public, or at least `BluetoothManager`. The bindings stuff feels a bit like "implementation details" to me: we decided to abstract JNI this way, but it's not necessarily the way the user would want to. Or if they're just getting the pointer from java and passing it to bluest they might not even need any abstractions, just jni-sys.

keeping the bindings private allows upgrading jni-glue/jni-bindgen to new versions that break absolutely everything, without breaking bluest's public api.

I don't like using raw pointers and unsafe either, but given the "generate ad-hoc bindings and embed them in the crate" it makes more sense I think. If we were using `jni-andtoid-sys` then it'd make nmore sense to take a `Global<BluetoothManager>` indeed.

Reply:
Ah, ok. Makes sense.